compiler/rustc_hir_typeck/src/demand.rs

```diff
@@ -205,6 +210,157 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         (expected, Some(err))
     }
 
+    fn point_inference_types(&self, err: &mut Diagnostic, expr: &hir::Expr<'_>) -> bool {
+        let tcx = self.tcx;
+        let map = self.tcx.hir();
+
+        // Hack to make equality checks on types with inference variables and regions useful.
```

Comment:
This is so that I can `ty == ty`. It's like if `can_eq` *didn't* coerce `_` and `{integer}`, so I can compare `Ty<_r#1>` with `Ty<_r#2>` as equal, but `Ty<_>` with `Ty<X>` are unequal. In effect this is the same as doing ty.to_string() == ty.to_string() (except when `-Zverbose` is set).

Reply:
Is this *just* for region variables? I'm not sure if I understand why we would want to treat *type* (or const) inference variables as equal?

Reply:
No, for region variables I can just `erase_regions`, this is for type inference variables. I want to be able to say "are these two types the same *including* inference variables?" As in `Ty<_> == Ty<_>` (regardless of what `_` is, they can have different `TyVid`), but also `Ty<_> != Ty<()>`.